pkg/webrtcclient/types.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package webrtcclient
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/pion/webrtc/v3"
+)
+
+const (
+	RequestOfferMsgType string = "request-offer"
+	ICECandidateMsgType string = "ice-candidate"
+	OfferMsgType        string = "offer"
+	AnswerMsgType       string = "answer"
+)
+
+type RequestOfferMsg struct {
+	Type       string             `json:"type"`
+	ICEServers []webrtc.ICEServer `json:"ice_servers,omitempty`
+}
+
+func NewRequestOfferMsg(servers []webrtc.ICEServer) *RequestOfferMsg {
+	return &RequestOfferMsg{
+		Type:       RequestOfferMsgType,
+		ICEServers: servers,
+	}
+}
+
+type ICECandidateMsg struct {
+	Type      string                  `json:"type"`
+	Candidate webrtc.ICECandidateInit `json:"candidate"`
+}
+
+func NewICECandidateMsg(candidate webrtc.ICECandidateInit) *ICECandidateMsg {
+	return &ICECandidateMsg{
+		ICECandidateMsgType,
+		candidate,
+	}
+}
+
+// The offer and answer messages have the same structure as the
+// webrtc.SessionDescription so they don't to be redefined here.
+
+// Allows converting a generic JSON object representation, like
+// map[string]interface{}, to a specific class such as ICECandidateMsg.
+func Reshape(original interface{}, reshaped interface{}) error {
```

Comment:
nit: Go 1.18 introduced generics which could make this function easier to read.

Reply:
I was thinking on avoiding the second parameter and returning: `(*K, error) instead. 

On a different note, can we specify original as map[string]interface{}?

Reply:
Refactored to use generics.

Can't make original be `map` because it's used on objects received from the server, which could be arrays too.